lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -142,32 +154,86 @@ defmodule ExDoc.Formatter.HTML.Autolink do
   end
 
   defp typespec_to_string(ast, typespecs, aliases, lib_dirs) do
-    Macro.to_string(ast, fn
-      {name, _, args}, string when is_atom(name) and is_list(args) ->
-        arity = length(args)
-        if {name, arity} in typespecs do
-          n = enc_h("#{name}")
-          {string_to_link, string_with_parens} = split_string_to_link(string)
-          ~s[<a href="#t:#{n}/#{arity}">#{h(string_to_link)}</a>#{string_with_parens}]
-        else
-          string
-        end
-      {{:., _, [alias, name]}, _, args}, string when is_atom(name) and is_list(args) ->
-        alias = expand_alias(alias)
-        if source = get_source(alias, aliases, lib_dirs) do
-          n = enc_h("#{name}")
-          {string_to_link, string_with_parens} = split_string_to_link(string)
-          ~s[<a href="#{source}#{enc_h(inspect alias)}.html#t:#{n}/#{length(args)}">#{h(string_to_link)}</a>#{string_with_parens}]
-        else
-          string
-        end
-      _, string ->
-        string
-    end)
+    ref = make_ref()
+
+    {ast, placeholders} =
+      Macro.prewalk(ast, %{}, fn
+        {:::, _, [{name, meta, args}, right]}, placeholders when is_atom(name) and is_list(args) ->
+          {{:::, [], [{{ref, name}, meta, args}, right]}, placeholders}
+
+        # Consume this form so that we don't autolink `foo` in `foo :: bar`
+        {{^ref, name}, _, args}, placeholders when is_atom(name) and is_list(args) ->
+          {{name, [], args}, placeholders}
+
+        {name, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
+          arity = length(args)
+
+          if {name, arity} in typespecs do
+            string = Macro.to_string(form)
+            n = enc_h("#{name}")
+            {string_to_link, _string_with_parens} = split_string_to_link(string)
+            string = ~s[<a href="#t:#{n}/#{arity}">#{h(string_to_link)}</a>]
+
+            put_placeholder(form, string, placeholders)
+          else
+            {form, placeholders}
+          end
+
+        {{:., _, [alias, name]}, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
+          alias = expand_alias(alias)
+
+          if source = get_source(alias, aliases, lib_dirs) do
+            string = Macro.to_string(form)
+            n = enc_h("#{name}")
+            {string_to_link, _string_with_parens} = split_string_to_link(string)
+            string = ~s[<a href="#{source}#{enc_h(inspect alias)}.html#t:#{n}/#{length(args)}">#{h(string_to_link)}</a>]
+
+            put_placeholder(form, string, placeholders)
+          else
+            {form, placeholders}
+          end
+
+        form, placeholders ->
+          {form, placeholders}
+      end)
+
+    ast
+    |> format_ast()
+    |> replace_placeholders(placeholders)
   end
 
-  defp short_typespec?(ast) do
-    byte_size(Macro.to_string(ast)) <= 70
+  defp put_placeholder(form, string, placeholders) do
+    id = zero_pad(map_size(placeholders), 3)
```

Comment:
I don't think we need to pad, no?

Reply:
yup, not anymore as there's always a trailing `_`